Add TFM and version support query to marshaller interface.

AaronRobinsonMSFT opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/60662

If any marshaller fails to support the current tfm version, fallback to using the built-in `DllImport` attribute.

This logic permits using the source generator for any .NET TFM and will gracefully "fallback" to only support platforms that can consume the generated source.

/cc @jkoritzinsky @elinor-fung @stephentoub @eerhardt

AaronRobinsonMSFT committed:
Add TFM and version support query to marshaller interface.

If any marshaller fails to support the current tfm version, fallback
to using the built-in DllImport attribute.

jkoritzinsky reviewed (commented):
The approach for checking supportability based on TFM works great for any out-of-band projects and any external customers, but still might cause issues with solving the following item on our list of features to productize:

> Enable DllImportGenerator by default for all source projects instead of just NETCoreApp with specific references (closer to an in-box generator experience)

One option to solve this problem would be to just say "we'll enable the generator for all source projects and if someone adds a stub that needs a missing type, we'll the dotnet/runtime dev will see an error message about a type that's not available and they'll have to add a reference themselves". Given how rare this will likely be (maybe once in a release cycle max), I think this could be a reasonable solution if we add something in the docs about it.

jkoritzinsky commented:
The problem would be when we're in the product since assemblies in the product reference a subset of the ref assemblies. For example, we added a direct reference to `System.Runtime.CompilerServices.Unsafe` to `System.Console` in https://github.com/dotnet/runtime/pull/59579 to ensure that the `Unsafe` class was available for use.

As of today, every assembly with P/Invokes in dotnet/runtime references the assemblies needed, but that isn't guaranteed to be the case if someone adds a new assembly in the repo.

jkoritzinsky reviewed (approved):
As we discussed offline, my concerns are for a small enough group (a small subset of dotnet/runtime contributors) that we can add some in-repo documentation to handle this case when we update our dotnet/runtime interop best practices/guidance.

elinor-fung commented:
I think we should add:
- doc about the target framework support/fallback behaviour - new section in the compatibility doc? That doc currently says we require .NET 6+.
- unit tests for down-level targets
  - signature that should be supported to validate we generate a stub
  - signature that should not be supported to validate we generate a forwarder

AaronRobinsonMSFT committed:
Add tests for fallback generation on downlevel or generation if marshallers
  are supported on the target framework.

Update compatibility document with fallback support.

elinor-fung commented on src/libraries/System.Runtime.InteropServices/tests/DllImportGenerator.UnitTests/Compiles.cs:
```diff
@@ -310,27 +310,38 @@ public async Task ValidateSnippetsWithPreprocessorDefintions(string source, IEnu
 
             var newComp = TestUtils.RunGenerators(
                 comp,
-                new DllImportGeneratorOptionsProvider(useMarshalType: false, generateForwarders: true),
```
You gon rid of the test for the `GenerateForwarders` option?

AaronRobinsonMSFT replied:
Yeah. Is this something we want to validate?

elinor-fung commented on src/libraries/System.Runtime.InteropServices/tests/DllImportGenerator.UnitTests/TestUtils.cs:
```diff
@@ -89,37 +123,54 @@ public static async Task<Compilation> CreateCompilation(SyntaxTree[] sources, Ou
         /// <param name="allowUnsafe">Whether or not use of the unsafe keyword should be allowed</param>
         /// <returns>The resulting compilation</returns>
-        public static Task<Compilation> CreateCompilationWithReferenceAssemblies(string source, ReferenceAssemblies referenceAssemblies, OutputKind outputKind = OutputKind.DynamicallyLinkedLibrary, bool allowUnsafe = true)
+        public static async Task<Compilation> CreateCompilationWithReferenceAssemblies(string source, TestTargetFramework targetFramework, OutputKind outputKind = OutputKind.DynamicallyLinkedLibrary, bool allowUnsafe = true)
```
`CreateCompilationForTargetFramework`?

elinor-fung replied:
Actually, maybe merge this with `CreateCompilation` now with a `targetFramework` default of `TestTargetFramework.Net`?

AaronRobinsonMSFT replied:
Done.

elinor-fung commented on docs/design/libraries/DllImportGenerator/Compatibility.md:
```diff
@@ -4,7 +4,11 @@ Documentation on compatibility guidance and the current state. The version headi
+
+### Fallback mechanism
+
+In the event a marshaller would generate code that has a specific TFM requirement, the generator will instead produce a normal `DllImportAttribute` declaration. This fallback mechanism enables the use of `GeneratedDllImportAttribute` in most circumstances and permits the conversion from `DllImportAttribute` to `GeneratedDllImportAttribute` to be across most code bases. There are instances where t ...
```
```suggestion
In the event a marshaller would generate code that has a specific target framework or version requirement that is not satisfied, the generator will instead produce a normal `DllImportAttribute` declaration. This fallback mechanism enables the use of `GeneratedDllImportAttribute` in most circumstances and permits the conversion from `DllImportAttribute` to `GeneratedDllImportAttribute` to be across most code bases. There are instances where the generator will not be able to handle signatures or configuration. For example, uses of `StringBuilder` are not supported in any form and consumers should retain uses of `DllImportAttribute`. Additionally, `GeneratedDllImportAttribute` cannot represent all settings available on `DllImportAttribute`&mdash;see below for details.
```

Just so it doesn't sound like the generator and marshallers are looking at the actual moniker.

AaronRobinsonMSFT committed:
Update docs/design/libraries/DllImportGenerator/Compatibility.md

Co-authored-by: Elinor Fung <elfung@microsoft.com>

AaronRobinsonMSFT committed:
Convert CreateCompilationWithReferenceAssemblies to CreateCompilation overload.
